include/rcutils/testing/fault_injection.h

```diff
@@ -0,0 +1,134 @@
+ * `RCUTILS_FAULT_INJECTION_MAYBE_RETURN_ERROR` will not cause an early return.
+ *
+ * This macro is thread-safe, and ensures that at most one invocation results in a failure for each
+ * time the fault injection counter is set with `RCUTILS_FAULT_INJECTION_SET_COUNT`
+ *
+ * \param return_value_on_error the value to return in the case of fault injected failure.
+ */
+#define RCUTILS_FAULT_INJECTION_MAYBE_RETURN_ERROR(return_value_on_error) \
+  if (RCUTILS_FAULT_INJECTION_FAIL_NOW == _rcutils_fault_injection_maybe_fail()) { \
+    printf( \
+      "%s:%d Injecting fault and returning " #return_value_on_error "\n", __FILE__, __LINE__); \
+    return return_value_on_error; \
+  }
+
+/**
+ * \def RCUTILS_FAULT_INJECTION_SET_COUNT
+ * \brief Atomically set the fault injection counter.
+ *
+ * There will be at most one fault injected failure per call to RCUTILS_FAULT_INJECTION_SET_COUNT.
+ * To test all reachable fault injection locations, call this macro inside a loop and set the count
+ * to an incrementing count variable.
+ *
+ *  for (int i = 0; i < SUFFICIENTLY_LARGE_ITERATION_COUNT; ++i) {
+ *    RCUTILS_FAULT_INJECTION_SET_COUNT(i);
+ *    ... // Call function under test
+ *  }
+ * ASSERT_LT(RCUTILS_FAULT_INJECTION_NEVER_FAIL, RCUTILS_FAULT_INJECTION_GET_COUNT());
+ *
+ * Where SUFFICIENTLY_LARGE_ITERATION_COUNT is a value larger than the maximum expected calls to
+ * `RCUTILS_FAULT_INJECTION_MAYBE_RETURN_ERROR`. This last assertion just insures that your choice for
+ * SUFFICIENTLY_LARGE_ITERATION_COUNT was large enough. To avoid having to choose this count
+ * yourself, you can use a do-while loop.
+ *
+ * int i = 0;
+ * do {
+ *   RCUTILS_FAULT_INJECTION_SET_COUNT(i++);
+ *    ... // Call function under test
+ * } while (RCUTILS_FAULT_INJECTION_GET_COUNT() <= RCUTILS_FAULT_INJECTION_NEVER_FAIL);
+ *
+ * \param count The count to set the fault injection counter to. If count is negative, then fault
+ * injection errors will be disabled. The counter is globally initialized to
+ * RCUTILS_FAULT_INJECTION_NEVER_FAIL.
+ */
+#define RCUTILS_FAULT_INJECTION_SET_COUNT(count) \
+  _rcutils_fault_injection_set_count(count);
+
+/**
+ * \def RCUTILS_FAULT_INJECTION_GET_COUNT
+ * \brief Atomically get the fault injection counter value
+ *
+ * Use this macro after running the code under test to check whether the counter reached a negative
+ * value. This is helpful so you can verify that you ran the fault injection test in a loop a
+ * sufficient number of times. Likewise, if the code under test returned with an error, but the
+ * count value was greater or equal to 0, then the failure was not caused by the fault injection
+ * counter.
+ */
+#define RCUTILS_FAULT_INJECTION_GET_COUNT() \
+  _rcutils_fault_injection_get_count();
+
+#define RCUTILS_FAULT_INJECTION_TEST(code) \
+  do { \
+    int fault_injection_count = 0; \
+    do { \
+      RCUTILS_FAULT_INJECTION_SET_COUNT(fault_injection_count++); \
+      code; \
```

Comment:
@brawner nice! I still think we need a way to disable fault injection temporarily within the `code` statement, otherwise you may hit a fault while handling or checking the outcome of the piece of functionality that is actually under test.

Reply:
That can be done by way of: 
```
int current_count = RCUTILS_FAULT_INJECTION_GET_COUNT();
RCUTILS_FAULT_INJECTION_SET_COUNT(RCUTILS_FAULT_INJECTION_NEVER_FAIL);
...
RCUTILS_FAULT_INJECTION_SET_COUNT(current_count);
```

Do you think we need something more like:

```
RCUTILS_FAULT_INJECTION_PAUSE_TEST();
...
RCUTILS_FAULT_INJECTION_RESUME_TEST();
```

Reply:
Yeah, some sugar would be nice. I'd personally like to have something along the lines of: 

```c++
RCUTILS_FAULT_INJECTION_NO_INJECT({
  // ...
});
```